frontend/ui/widget/iconbutton.lua

```diff
@@ -158,4 +158,11 @@ function IconButton:onTapSelect()
     self:onTapIconButton()
 end
 
+function IconButton:setIcon(icon)
+    if icon ~= self.icon then
+        self.icon = icon
+        self:init()
```

Comment:
It's possible you need to do `self:free()` before `self:init()` to clean the previous instance bb and avoid some micro memory leak. (I'd say: try adding it, and if it doesn't crash, we're good :)
(As IconButton extends ImageWidget, this will call ImageWidget:free() which seem to do enough checks to do it only if it's right to free it.)

Reply:
I took it from
https://github.com/koreader/koreader/blob/a703b213f76e660be60f2b65601967f0f81c3410/frontend/ui/widget/button.lua#L209-L215
Check there too?

Reply:
Yes, I think it's needed there too.
We do `self.label_widget:free()` in the `self.avoid_text_truncation` check when forgetting the TextWidget to use a TextBoxWidget.
So, I think in the 2 places we call self:init() (including the :setText() one), we should do:
`self.label_widget:free()`
It seems to be either a Text*Widget or an IconWidget, which should all have a .free() method, but to be super sure, we could do:
```lua
if self.label_widget and self.label_widget.free then
  self.label_widget:free()
end
```